File: ios/VineTrack/App/Reporting/SeasonWindow.swift
import Foundation

/// A single production season expressed as a **date range**, derived from the
/// vineyard's shared season-start setting (SQL 108 `season_start_month/day`).
///
/// This is the reporting counterpart to `VintageResolver`. Where the resolver
/// answers "which vintage does this date belong to?", `SeasonWindow` answers
/// the inverse — "which dates belong to this vintage?" — so ordinary dated
/// operational records (sprays, trips, fuel, maintenance, fertiliser, growth
/// stages) can be filtered by their EXISTING event-date column with no stored
/// vintage of their own.
///
/// Stored vintage remains reserved for records inherently owned by a vintage:
/// Yield Estimates, Grape Allocations, Picking/Yield records and Damage.
///
/// Boundaries are vineyard-local: the window is built in the vineyard's
/// configured timezone (`AppSettings.resolvedTimeZone`), not the device's, so
/// an operator travelling interstate sees the same season as their colleague
/// standing in the block.
///
/// The range is half-open — `start ..< endExclusive` — which is the only form
/// that is correct for `timestamptz` event columns. An inclusive end-of-day
/// comparison silently drops records stamped in the final second of the season.
nonisolated struct SeasonWindow: Equatable, Identifiable, Sendable {

    /// Vintage year — the calendar year in which the season ENDS.
    let vintage: Int

    /// First instant of the season, vineyard-local.
    let start: Date

    /// First instant of the FOLLOWING season. Exclusive upper bound.
    let endExclusive: Date

    var id: Int { vintage }

    /// Last representable instant inside the season, for display only.
    /// Never use this for filtering — filter with `contains(_:)`.
    var displayEnd: Date { endExclusive.addingTimeInterval(-1) }

    /// True when `date` falls inside this season.
    func contains(_ date: Date) -> Bool {
        date >= start && date < endExclusive
    }

    /// True when `date` is inside the season, treating a missing date as
    /// excluded. Records with no reliable event date are reported separately
    /// rather than being silently dropped into the current season.
    func contains(optional date: Date?) -> Bool {
        guard let date else { return false }
        return contains(date)
    }

    // MARK: - Construction

    /// Number of previous seasons offered by the selector alongside the current one.
    static let previousSeasonCount = 15

    /// The window for a specific `vintage`.
    static func window(
        vintage: Int,
        seasonStartMonth: Int,
        seasonStartDay: Int,
        timeZone: TimeZone
    ) -> SeasonWindow {
        var calendar = Calendar(identifier: .gregorian)
        calendar.timeZone = timeZone

        let month = min(max(seasonStartMonth, 1), 12)
        let day = max(seasonStartDay, 1)

        // A 1 January season is contained within a single calendar year, so the
        // vintage IS that year. Every other season start straddles the new year,
        // so the season opens in the PREVIOUS calendar year. This mirrors
        // `VintageResolver.vintageYear` exactly — the two must never disagree,
        // or a record would be filtered out of the very season it displays as.
        let startYear = (month == 1 && day == 1) ? vintage : vintage - 1

        let start = seasonStart(inYear: startYear, month: month, day: day, calendar: calendar)
        let end = seasonStart(inYear: startYear + 1, month: month, day: day, calendar: calendar)
        return SeasonWindow(vintage: vintage, start: start, endExclusive: end)
    }

    /// The window containing `date` — i.e. the season that date belongs to.
    static func window(
        containing date: Date,
        seasonStartMonth: Int,
        seasonStartDay: Int,
        timeZone: TimeZone
    ) -> SeasonWindow {
        var calendar = Calendar(identifier: .gregorian)
        calendar.timeZone = timeZone
        let vintage = VintageResolver.vintageYear(
            for: date,
            seasonStartMonth: seasonStartMonth,
            seasonStartDay: seasonStartDay,
            calendar: calendar
        )
        return window(
            vintage: vintage,
            seasonStartMonth: seasonStartMonth,
            seasonStartDay: seasonStartDay,
            timeZone: timeZone
        )
    }

    /// The vintage in progress right now, vineyard-local.
    static func currentVintage(
        seasonStartMonth: Int,
        seasonStartDay: Int,
        timeZone: TimeZone,
        now: Date = Date()
    ) -> Int {
        var calendar = Calendar(identifier: .gregorian)
        calendar.timeZone = timeZone
        return VintageResolver.vintageYear(
            for: now,
            seasonStartMonth: seasonStartMonth,
            seasonStartDay: seasonStartDay,
            calendar: calendar
        )
    }

    /// Descending vintage options for a season selector: the current season
    /// plus the previous 15.
    ///
    /// `selected` is always included even when it falls outside that window, so
    /// a deep link into an old record can never land on a missing option. An
    /// `earliestRecordVintage` extends the list backwards only as far as data
    /// actually exists, so a new vineyard isn't offered 15 empty seasons.
    static func availableVintages(
        currentVintage: Int,
        selected: Int? = nil,
        earliestRecordVintage: Int? = nil
    ) -> [Int] {
        var years = Set<Int>([currentVintage])
        let floor = max(
            currentVintage - previousSeasonCount,
            earliestRecordVintage ?? (currentVintage - previousSeasonCount)
        )
        if floor <= currentVintage {
            for year in floor...currentVintage { years.insert(year) }
        }
        if let selected { years.insert(selected) }
        return years.filter { $0 > 0 }.sorted(by: >)
    }

    /// "2027 · Current" for the season in progress, otherwise "2027".
    static func label(for vintage: Int, currentVintage: Int) -> String {
        vintage == currentVintage ? "\(String(vintage)) · Current" : String(vintage)
    }

    /// Season start in `year`, with the configured day clamped to the month's
    /// real length so a 29 February start behaves as 28 February in non-leap
    /// years — identical to the server resolver.
    private static func seasonStart(inYear year: Int, month: Int, day: Int, calendar: Calendar) -> Date {
        var components = DateComponents()
        components.year = year
        components.month = month
        components.day = 1
        guard let firstOfMonth = calendar.date(from: components) else {
            return Date(timeIntervalSince1970: 0)
        }
        let maxDay = calendar.range(of: .day, in: .month, for: firstOfMonth)?.count ?? 28
        components.day = min(day, maxDay)
        guard let start = calendar.date(from: components) else { return firstOfMonth }
        return calendar.startOfDay(for: start)
    }
}

extension AppSettings {
    /// Vintage in progress for this vineyard, using its shared season start and
    /// its own timezone.
    nonisolated var currentSeasonVintage: Int {
        SeasonWindow.currentVintage(
            seasonStartMonth: seasonStartMonth,
            seasonStartDay: seasonStartDay,
            timeZone: resolvedTimeZone
        )
    }

    /// Season window for `vintage` under this vineyard's settings.
    nonisolated func seasonWindow(for vintage: Int) -> SeasonWindow {
        SeasonWindow.window(
            vintage: vintage,
            seasonStartMonth: seasonStartMonth,
            seasonStartDay: seasonStartDay,
            timeZone: resolvedTimeZone
        )
    }

    /// Season window containing `date` under this vineyard's settings.
    nonisolated func seasonWindow(containing date: Date) -> SeasonWindow {
        SeasonWindow.window(
            containing: date,
            seasonStartMonth: seasonStartMonth,
            seasonStartDay: seasonStartDay,
            timeZone: resolvedTimeZone
        )
    }
}

File: android-vinetrack/app/src/main/java/com/rork/vinetrack/data/SeasonWindow.kt
package com.rork.vinetrack.data

import java.time.Instant
import java.time.LocalDate
import java.time.YearMonth
import java.time.ZoneId

/**
 * A single production season expressed as a **date range**, derived from the
 * vineyard's shared season-start setting (sql/108 `season_start_month/day`).
 *
 * Reporting counterpart to [VintageResolver]: where the resolver answers "which
 * vintage does this date belong to?", `SeasonWindow` answers the inverse —
 * "which dates belong to this vintage?" — so ordinary dated operational records
 * (sprays, trips, fuel, maintenance, fertiliser, growth stages) can be filtered
 * by their EXISTING event-date column with no stored vintage of their own.
 *
 * Stored vintage stays reserved for records inherently owned by a vintage:
 * Yield Estimates, Grape Allocations, Picking/Yield records and Damage.
 *
 * The range is half-open — `start until endExclusive` — the only form that is
 * correct for timestamp columns. An inclusive end-of-day bound silently drops
 * records stamped in the final second of the season.
 *
 * Mirrors the iOS `SeasonWindow`; the two must not diverge.
 */
data class SeasonWindow(
    /** Vintage year — the calendar year in which the season ENDS. */
    val vintage: Int,
    /** First day of the season, vineyard-local. */
    val start: LocalDate,
    /** First day of the FOLLOWING season. Exclusive upper bound. */
    val endExclusive: LocalDate,
) {
    /** Last day inside the season — for display only, never for filtering. */
    val displayEnd: LocalDate get() = endExclusive.minusDays(1)

    /** First instant of the season in [zone]. */
    fun startEpochMs(zone: ZoneId = ZoneId.systemDefault()): Long =
        start.atStartOfDay(zone).toInstant().toEpochMilli()

    /** First instant of the following season in [zone] — exclusive. */
    fun endExclusiveEpochMs(zone: ZoneId = ZoneId.systemDefault()): Long =
        endExclusive.atStartOfDay(zone).toInstant().toEpochMilli()

    /** True when [date] falls inside this season. */
    fun contains(date: LocalDate): Boolean =
        !date.isBefore(start) && date.isBefore(endExclusive)

    /**
     * True when [epochMs] falls inside this season, evaluated in [zone].
     * A null timestamp is excluded rather than being silently swept into the
     * current season — records with no reliable event date are surfaced
     * separately.
     */
    fun containsEpochMs(epochMs: Long?, zone: ZoneId = ZoneId.systemDefault()): Boolean {
        if (epochMs == null) return false
        val date = Instant.ofEpochMilli(epochMs).atZone(zone).toLocalDate()
        return contains(date)
    }

    /** True when a `yyyy-MM-dd` (or ISO timestamp) string falls inside this season. */
    fun containsIsoDate(iso: String?): Boolean {
        val date = parseIsoDate(iso) ?: return false
        return contains(date)
    }

    companion object {
        /** Previous seasons offered by a selector alongside the current one. */
        const val PREVIOUS_SEASON_COUNT = 15

        /** The window for a specific [vintage]. */
        fun forVintage(vintage: Int, seasonStartMonth: Int, seasonStartDay: Int): SeasonWindow {
            val month = seasonStartMonth.coerceIn(1, 12)
            val day = seasonStartDay.coerceAtLeast(1)

            // A 1 January season sits inside a single calendar year, so the
            // vintage IS that year. Every other start straddles the new year, so
            // the season opens in the PREVIOUS calendar year. This mirrors
            // VintageResolver.vintageYear exactly — if the two disagreed, a
            // record could be filtered out of the season it displays as.
            val startYear = if (month == 1 && day == 1) vintage else vintage - 1
            return SeasonWindow(
                vintage = vintage,
                start = seasonStart(startYear, month, day),
                endExclusive = seasonStart(startYear + 1, month, day),
            )
        }

        /** The season containing [date]. */
        fun containing(date: LocalDate, seasonStartMonth: Int, seasonStartDay: Int): SeasonWindow =
            forVintage(
                VintageResolver.vintageYear(date, seasonStartMonth, seasonStartDay),
                seasonStartMonth,
                seasonStartDay,
            )

        /** The vintage in progress right now. */
        fun currentVintage(
            seasonStartMonth: Int,
            seasonStartDay: Int,
            today: LocalDate = LocalDate.now(),
        ): Int = VintageResolver.vintageYear(today, seasonStartMonth, seasonStartDay)

        /**
         * Descending vintage options: the current season plus the previous 15.
         *
         * [selected] is always included even when outside that window, so a deep
         * link into an old record can never land on a missing option.
         * [earliestRecordVintage] trims the list to seasons that actually hold
         * data, so a young vineyard isn't offered 15 empty seasons.
         */
        fun availableVintages(
            currentVintage: Int,
            selected: Int? = null,
            earliestRecordVintage: Int? = null,
        ): List<Int> {
            val floor = maxOf(
                currentVintage - PREVIOUS_SEASON_COUNT,
                earliestRecordVintage ?: (currentVintage - PREVIOUS_SEASON_COUNT),
            )
            val years = sortedSetOf<Int>()
            if (floor <= currentVintage) for (year in floor..currentVintage) years.add(year)
            years.add(currentVintage)
            selected?.let { years.add(it) }
            return years.filter { it > 0 }.sortedDescending()
        }

        /** "2027 · Current" for the season in progress, otherwise "2027". */
        fun label(vintage: Int, currentVintage: Int): String =
            if (vintage == currentVintage) "$vintage · Current" else "$vintage"

        /**
         * Season start in [year], with the configured day clamped to the month's
         * real length so a 29 February start behaves as 28 February in non-leap
         * years — identical to the server resolver.
         */
        private fun seasonStart(year: Int, month: Int, day: Int): LocalDate {
            val yearMonth = YearMonth.of(year, month)
            return yearMonth.atDay(minOf(day, yearMonth.lengthOfMonth()))
        }

        /** Parses `yyyy-MM-dd` or a full ISO timestamp; null when unusable. */
        fun parseIsoDate(iso: String?): LocalDate? {
            val raw = iso?.trim().orEmpty()
            if (raw.isEmpty()) return null
            runCatching { return LocalDate.parse(raw.take(10)) }
            return runCatching { Instant.parse(raw).atZone(ZoneId.systemDefault()).toLocalDate() }.getOrNull()
        }
    }
}
